# Kapitel 4 - Koduppgift 11

In [1]:
from sklearn.metrics import classification_report

# y_true innehåller de faktiska (sanna) klasserna för 5 observationer
# Klasserna är: 0, 1, 2, 2, 2
y_true = [0, 1, 2, 2, 2]

# y_pred innehåller modellens predicerade klasser för samma 5 observationer
# Modellen predicerade: 0, 0, 2, 2, 1
y_pred = [0, 0, 2, 2, 1]

# Skriv ut classification_report
# Parametern target_names är valfri och används för att ge namn åt klasserna
# Här är target_names = ['class 0', 'class 1', 'class 2']
print(classification_report(y_true, y_pred, target_names=['class 0', 'class 1', 'class 2']))

              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5



##### Koden används för att kontrollera hur bra en klassificeringsmodell fungerar genom att jämföra vad modellen tror (prediktioner). Modellens träffsäkerhet (accuracy) är 60%, vilket innebär att den svarade rätt i 3 av 5 fall. 

##### Klass 0: Modellen lyckas hitta det enda riktiga fallet av klass 0 (därför är recall = 1.0). Däremot gissar den klass 0 en gång för mycket, vilket gör att precisionen bara blir 0.50. Det betyder att den ibland “tror” något är klass 0 fast det inte är det.

##### Klass 1: Här misslyckas modellen helt. Den hittar inte det enda riktiga fallet och gör dessutom en felaktig gissning. Därför blir alla mått (precision, recall, F1-score) lika med 0.

##### Klass 2: Det finns tre riktiga fall av klass 2, och modellen hittar två av dem (recall = 0.67). Samtidigt är alla dess gissningar på klass 2 korrekta (precision = 1.0). Det betyder att modellen är säker när den väl väljer klass 2, men den missar ibland att upptäcka den.

##### Macro average: Ett enkelt medelvärde där alla klasser räknas lika mycket, oavsett hur många exempel de har. 
##### Weighted average: Ett medelvärde där klasser med fler exempel väger tyngre och påverkar resultatet mer.
 
